src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V66/Eth66ProtocolHandler.cs

```diff
@@ -276,12 +278,51 @@ protected override async Task<TxReceipt[][]> SendRequest(V63.Messages.GetReceipt
             }
 
             GetReceiptsMessage msg66 = new() { EthMessage = message };
-            return await SendRequestGeneric(
+            return await SendRequestGenericEth66(
                 _receiptsRequests66,
                 msg66,
                 TransferSpeedType.Receipts,
                 static (_) => $"{nameof(GetReceiptsMessage)}",
                 token);
         }
+
+        private async Task<TResponse> SendRequestGenericEth66<T66, TRequest, TResponse>(
+            MessageDictionary<T66, TRequest, TResponse> messageQueue,
+            T66 message,
+            TransferSpeedType speedType,
+            Func<T66, string> describeRequestFunc,
+            CancellationToken token
+        )
+            where T66 : Eth66Message<TRequest>
+            where TRequest : P2PMessage
+        {
+            message.RequestId = Interlocked.Increment(ref _currentRequestId);
+            Request<T66, TResponse> request = new(message);
+            messageQueue.Send(request);
+
+            Task<TResponse> task = request.CompletionSource.Task;
+            using CancellationTokenSource delayCancellation = new();
+            using CancellationTokenSource compositeCancellation =
+                CancellationTokenSource.CreateLinkedTokenSource(token, delayCancellation.Token);
```

Comment:
It is still sending requests and waiting for responses (creating `task = request.CompletionSource.Task` and awaiting it). When we send request we are adding `RequestId` and `Request` itself to `MessageDictionary`. We should consider sending requests (and having `Ids` and `Requests` in dictionary), and when we receive message (potential response), we can check in dictionary if we requested it (by `Id`) and act accordingly (pair `request`<->`response` or throw `subprotocol exception` if message was not requested). I think we can improve performance a lot if we will be able to drop awaiting here. I failed when I was trying to introduce it. Do you think it is doable @LukaszRozmej ?

Reply:
This function is async, so the execution is async. I won't worry too much about that. I doubt we will improve performance if any because.
1. We still wait for response (processing of response) in sync dispatcher before dispatching another request instead of dispatching multiple request per peer at the same time.
2. For snap sync, the limit is the get address range call which is actually sequential. We can try to request next address while processing current address range, but that skip validation and such.
3. For old block/receipts, if you are not network or io limited, the deserialization of all P2P message is capped to a single thread, which is a problem because there are Keccak call in the deserialization method.
4. For forward sync, we straight up did not download from multiple peer at the same time.

Reply:
Hm... But then do we achieve anything with this PR? We didn't really change anything meaningful about the request handling, it is exactly the same?

I think this might be worth experimenting with in conjunction with other things you are mentioning. But truth be told not exactly sure.

Reply:
Don't know. Maybe the peer may respond out of order. Maybe something in rlpx that I don't realize.